packages/nimble-components/src/select/index.ts

```diff
@@ -969,39 +1007,90 @@ export class Select
+        const filter = this.filter.toLowerCase();
+        const normalizedFilter = diacriticInsensitiveStringNormalizer(filter);
+        return diacriticInsensitiveStringNormalizer(text).includes(
+            normalizedFilter
+        );
+    }
+
     /**
      * Filter available options by text value.
      *
      * @public
      */
     private filterOptions(): void {
-        const filter = this.filter.toLowerCase();
+        if (!this.$fastController.isConnected) {
+            return;
+        }
+
+        const filteredOptions: ListOption[] = [];
+        let lastVisibleElement: Element | undefined;
+        for (const element of this.slottedOptions) {
+            let elementHidden = false;
 
-        if (filter) {
-            this.filteredOptions = this.options.filter(option => {
-                const normalizedFilter = diacriticInsensitiveStringNormalizer(filter);
-                return (
-                    !option.hidden
-                    && diacriticInsensitiveStringNormalizer(option.text).includes(
-                        normalizedFilter
-                    )
+            if (element instanceof ListOptionGroup) {
+                if (element.hidden) {
+                    elementHidden = true;
+                    break; // no need to process hidden groups
+                }
+                const groupOptions = this.getGroupOptions(element);
+                const groupMatchesFilter = this.filterMatchesText(
+                    element.labelContent
                 );
-            });
-        } else {
-            this.filteredOptions = this.options.filter(
-                option => !option.hidden
-            );
+                let allOptionsHidden = true;
+                groupOptions.forEach(option => {
+                    option.visuallyHidden = groupMatchesFilter
+                        ? false
+                        : this.isOptionHidden(option);
+                    if (!option.visuallyHidden) {
+                        filteredOptions.push(option);
+                    }
+                    allOptionsHidden = option.visuallyHidden && allOptionsHidden;
+                });
+                element.visuallyHidden = allOptionsHidden || element.hidden;
+                elementHidden = element.visuallyHidden;
+            } else if (isNimbleListOption(element)) {
+                element.visuallyHidden = this.isOptionHidden(element);
+                elementHidden = element.visuallyHidden;
+                if (!element.visuallyHidden) {
+                    filteredOptions.push(element);
+                }
+            }
+
+            if (lastVisibleElement instanceof ListOptionGroup) {
+                lastVisibleElement.showBottomSeparator = true;
+            }
+
+            if (element instanceof ListOptionGroup && !elementHidden) {
+                element.showTopSeparator = lastVisibleElement instanceof ListOption;
+                element.showBottomSeparator = false;
+            }
```

Comment:
I think this whole block of code needs to be wrapped around a `!elementHidden` statement, rather than having that check being part of line 1073. As is, you will set `showBottomSeparator = true` on the last visible group, even if nothing is visible below it.

To see this in practice, you can look at your storybook page with filtering & grouping enabled. In the filter, type "1". Only the first group is shown, but it incorrectly has a bottom separator.